frontend-style.md

```diff
@@ -0,0 +1,351 @@
+# Frontend Style Guide
+
+## Structure
+
+## File organization
+
+All files should be organized by feature (e.g. math standard practice) instead of function (e.g. backbone collection). This includes:
+
+- Backbone models & collections
+- Components
+- Jest tests
+- Data mocks for jest tests
+
+Files pertaining to a single component specific to a feature should be organized as follows:
+
+```
+.
+└── my-page
+    ├── index.js
+    └── my-component
+        ├── index.js
+        ├── my-component.js
+        ├── my-component.scss
+        └── my-component.test.js
+```
+
+- `my-page/index.js` is a "handler" style component responsible for global
+  state management and API calls
+- `my-component/index.js` is a "container" style component responsible for
+  internal IO operations like user input or responding to timers
+- `my-component/my-component.js` is a pure, stateless "presentational" style
+  component
+- `my-component/my-component.test.js` contains jest tests for `my-component`
+- `my-component/my-component.scss` contains `scss` rules for `my-component`
+  that can be imported using `css modules`
+
+`my-page` should export a `React` class named `MyPage` to be referenced in the
+project's central `router`.
+
+## Formatting
+
+Use `prettier` for autoformatting. This way, we don't need to define a pedantic
+formatting style; we just let `prettier` decide for us.
+
+### JS Style Rules
+
+#### Variables
+
+Avoid variable mutation as much as possible.
+
+Use `const` by default.  If you must reassign a variable, use `let`. Never use
+`var`.
+
+```js
+// bad, name is not reassigned
+let name = "Doctor Brown"
+console.log({name}}
+
+// bad, var is outlawed
+var name = "Doctor Brown"
+console.log({name}}
+
+// good
+const name = "Doctor Brown"
+console.log({name})
+
+// fine, name is reassigned
+let name = "Doctor Brown"
+name = name.split(" ")
+console.log({name}}
+```
+
+#### Objects
+
+Use literal syntax to construct objects.
+
+```js
+// bad
+const item = new Object()
+
+// good
+const item = {}
+```
+
+Use property value shorthand when possible.
+
+```js
+// bad
+const user = { firstname: firstname }
+
+// good
+const user = { firstname }
+```
+
+Only quote properties that are invalid identifiers.
+
+```js
+// bad
+const user =
+  { "firstname": "Jeff"
+  , "age-group": "old"
+  }
+
+// good
+const user =
+  { firstname: "Jeff"
+  , "age-group": "old"
+  }
+```
+
+#### Arrays
+
+Use array literal syntax to construct arrays
+
+```js
+// bad
+const items = new Array();
+
+// good
+const items = []
+```
+
+## JSX Style Rules
+
+### Stateless Components
+
+Prefer stateless components when possible.
+
+```jsx
+// bad
+class Link extends React.Component<void, Props, void> {
+ render(): React.Element<any> {
+  return (
+    <a href="#foo">Link</a>
+  )
+ }
+}
+
+// good
+function Link(props: Props): React.Element<any> {
+  return <a href="#foo">Link</a>
+}
+```
```

Comment:
I am not sure if documenting things that we enforce via frontend tooling is valuable. This might mean creating multiple sources of the same information and might get out of sync.

Reply:
Yeah, I was just porting what we had in the confluence doc. All for nixing anything y'all don't find valuable